src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs

```diff
@@ -58,93 +64,126 @@ public virtual void CompileViews([NotNull] IBeforeCompileContext context)
             }
         }
 
-        protected virtual IReadOnlyList<RazorFileInfo> CreateCompilationDescriptors(
-                                                            [NotNull] IBeforeCompileContext context)
+        protected virtual IEnumerable<RazorFileInfo> CreateCompilationDescriptors(
+            [NotNull] IBeforeCompileContext context)
         {
-            var list = new List<RazorFileInfo>();
+            var filesToProcess = new List<RelativeFileInfo>();
+            GetFileInfosRecursive(root: string.Empty, razorFiles: filesToProcess);
 
-            foreach (var info in GetFileInfosRecursive(string.Empty))
-            {
-                var descriptor = ParseView(info, context);
+            var razorFiles = new RazorFileInfo[filesToProcess.Count];
+            var syntaxTrees = new SyntaxTree[filesToProcess.Count];
+            var parallelOptions = new ParallelOptions { MaxDegreeOfParallelism = MaxDegreesOfParallelism };
 
-                if (descriptor != null)
+            Parallel.For(0, filesToProcess.Count, parallelOptions, index =>
+            {
+                var file = filesToProcess[index];
+                var cacheEntry = PreCompilationCache.GetOrSet(file.RelativePath,
+                                                              file,
+                                                              OnCacheMiss);
+                if (cacheEntry != null)
                 {
-                    list.Add(descriptor);
+                    if (cacheEntry.Success)
+                    {
+                        syntaxTrees[index] = cacheEntry.SyntaxTree;
+                        razorFiles[index] = cacheEntry.FileInfo;
+                    }
+                    else
+                    {
+                        lock (context.Diagnostics)
```

Comment:
Having no \ very few diagnostic messages should be the more common case. Consequently a relatively low contention lock is preferable to using a concurrent collection. That said I couldn't necessarily see these in my profiles, so this might be a case of prematurely optimizing this bit. I'll use a separate lock for now - the code doesn't look any worse for having used the lock.